CA1508 False positive with char equality

### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.1 (Latest)

### Steps To Reproduce

````csharp
        private static bool TryParseUnit(string unit)
        {
            char last = unit[^1];
            if (last != 'b') // warning CA1508: 'last != 'b'' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
                return false;

            return true;
        }
````

### Expected behavior

No diagnostic as the condition is not a constant.

### Actual behavior

Report a diagnostic

### Additional data

No diagnostic is reported if I inline the variable

````
if (unit[^1] != 'b') // no diagnostic
````

--- a/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
+++ b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
@@ -153,10 +153,9 @@ namespace Microsoft.CodeQuality.Analyzers.Maintainability
                                         }
 
                                         var originalOperation = operationRoot.SemanticModel.GetOperation(operation.Syntax, operationBlockContext.CancellationToken);
-                                        if (originalOperation is IAssignmentOperation or
-                                            IVariableDeclaratorOperation or IInvocationOperation)
+                                        if (originalOperation.Parent is null || !ShouldAnalyze(originalOperation.Parent))
                                         {
-                                            // Skip compiler generated IsNull operation for assignment/variable declaration/invocation within a using.
+                                            // Skip compiler generated IsNull operation.
                                             continue;
                                         }
 

--- a/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
+++ b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
@@ -154,9 +154,9 @@ namespace Microsoft.CodeQuality.Analyzers.Maintainability
 
                                         var originalOperation = operationRoot.SemanticModel.GetOperation(operation.Syntax, operationBlockContext.CancellationToken);
                                         if (originalOperation is IAssignmentOperation or
-                                            IVariableDeclaratorOperation)
+                                            IVariableDeclaratorOperation or IInvocationOperation)
                                         {
-                                            // Skip compiler generated IsNull operation for assignment/variable declaration within a using.
+                                            // Skip compiler generated IsNull operation for assignment/variable declaration/invocation within a using.
                                             continue;
                                         }
 

--- a/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
+++ b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs
@@ -100,70 +100,72 @@ namespace Microsoft.CodeQuality.Analyzers.Maintainability
                                 continue;
                             }
 
-                            foreach (var operation in cfg.DescendantOperations())
+                            foreach (var block in cfg.Blocks)
                             {
-                                // Skip implicit operations.
-                                // However, 'IsNull' operations are compiler generated operations corresponding to
-                                // non-implicit conditional access operations, so we should not skip them.
-                                if (operation.IsImplicit && operation.Kind != OperationKind.IsNull)
+                                foreach (var operation in block.DescendantOperations())
                                 {
-                                    continue;
-                                }
+                                    // Skip implicit operations.
+                                    // However, 'IsNull' operations are compiler generated operations corresponding to
+                                    // non-implicit conditional access operations, so we should not skip them.
+                                    if (operation.IsImplicit && operation.Kind != OperationKind.IsNull)
+                                    {
+                                        continue;
+                                    }
 
-                                switch (operation.Kind)
-                                {
-                                    case OperationKind.BinaryOperator:
-                                        var binaryOperation = (IBinaryOperation)operation;
-                                        PredicateValueKind predicateKind = GetPredicateKind(binaryOperation);
-                                        if (predicateKind != PredicateValueKind.Unknown &&
-                                            (binaryOperation.LeftOperand is not IBinaryOperation leftBinary || GetPredicateKind(leftBinary) == PredicateValueKind.Unknown) &&
-                                            (binaryOperation.RightOperand is not IBinaryOperation rightBinary || GetPredicateKind(rightBinary) == PredicateValueKind.Unknown))
-                                        {
-                                            ReportAlwaysTrueFalseOrNullDiagnostic(operation, predicateKind);
-                                        }
+                                    switch (operation.Kind)
+                                    {
+                                        case OperationKind.BinaryOperator:
+                                            var binaryOperation = (IBinaryOperation)operation;
+                                            PredicateValueKind predicateKind = GetPredicateKind(binaryOperation);
+                                            if (predicateKind != PredicateValueKind.Unknown &&
+                                                (binaryOperation.LeftOperand is not IBinaryOperation leftBinary || GetPredicateKind(leftBinary) == PredicateValueKind.Unknown) &&
+                                                (binaryOperation.RightOperand is not IBinaryOperation rightBinary || GetPredicateKind(rightBinary) == PredicateValueKind.Unknown))
+                                            {
+                                                ReportAlwaysTrueFalseOrNullDiagnostic(operation, predicateKind);
+                                            }
 
-                                        break;
+                                            break;
 
-                                    case OperationKind.Invocation:
-                                    case OperationKind.IsPattern:
-                                        predicateKind = GetPredicateKind(operation);
-                                        if (predicateKind != PredicateValueKind.Unknown)
-                                        {
-                                            ReportAlwaysTrueFalseOrNullDiagnostic(operation, predicateKind);
-                                        }
+                                        case OperationKind.Invocation:
+                                        case OperationKind.IsPattern:
+                                            predicateKind = GetPredicateKind(operation);
+                                            if (predicateKind != PredicateValueKind.Unknown)
+                                            {
+                                                ReportAlwaysTrueFalseOrNullDiagnostic(operation, predicateKind);
+                                            }
 
-                                        break;
+                                            break;
 
-                                    case OperationKind.IsNull:
-                                        // '{0}' is always/never '{1}'. Remove or refactor the condition(s) to avoid dead code.
-                                        predicateKind = GetPredicateKind(operation);
-                                        DiagnosticDescriptor rule;
-                                        switch (predicateKind)
-                                        {
-                                            case PredicateValueKind.AlwaysTrue:
-                                                rule = AlwaysTrueFalseOrNullRule;
-                                                break;
+                                        case OperationKind.IsNull:
+                                            // '{0}' is always/never '{1}'. Remove or refactor the condition(s) to avoid dead code.
+                                            predicateKind = GetPredicateKind(operation);
+                                            DiagnosticDescriptor rule;
+                                            switch (predicateKind)
+                                            {
+                                                case PredicateValueKind.AlwaysTrue:
+                                                    rule = AlwaysTrueFalseOrNullRule;
+                                                    break;
 
-                                            case PredicateValueKind.AlwaysFalse:
-                                                rule = NeverNullRule;
-                                                break;
+                                                case PredicateValueKind.AlwaysFalse:
+                                                    rule = NeverNullRule;
+                                                    break;
 
-                                            default:
+                                                default:
+                                                    continue;
+                                            }
+
+                                            // Skip IsNull operations in compiler generated finally region for using statement
+                                            if (block.IsFirstBlockOfCompilerGeneratedFinally(cfg))
+                                            {
                                                 continue;
-                                        }
+                                            }
 
-                                        var originalOperation = operationRoot.SemanticModel.GetOperation(operation.Syntax, operationBlockContext.CancellationToken);
-                                        if (originalOperation.Parent is null || !ShouldAnalyze(originalOperation.Parent))
-                                        {
-                                            // Skip compiler generated IsNull operation.
-                                            continue;
-                                        }
-
-                                        var arg1 = operation.Syntax.ToString();
-                                        var arg2 = operation.Language == LanguageNames.VisualBasic ? "Nothing" : "null";
-                                        var diagnostic = operation.CreateDiagnostic(rule, arg1, arg2);
-                                        operationBlockContext.ReportDiagnostic(diagnostic);
-                                        break;
+                                            var arg1 = operation.Syntax.ToString();
+                                            var arg2 = operation.Language == LanguageNames.VisualBasic ? "Nothing" : "null";
+                                            var diagnostic = operation.CreateDiagnostic(rule, arg1, arg2);
+                                            operationBlockContext.ReportDiagnostic(diagnostic);
+                                            break;
+                                    }
                                 }
                             }
 

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -3360,6 +3360,45 @@ class B : A
 
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
+        [Fact, WorkItem(4411, "https://github.com/dotnet/roslyn-analyzers/issues/4411")]
+        public async Task NullCheck_AfterTryCast_04_NoDiagnostic()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+internal static class Class1
+{
+    public static ReadOnlyDictionary<TKey, ReadOnlyCollection<TValue>> AsReadOnlyItems<TKey, TValue>(this IDictionary<TKey, IList<TValue>> self, IEqualityComparer<TKey> keyComparer = null)
+    {
+        _ = self ?? throw new ArgumentNullException(nameof(self));
+
+        keyComparer ??= (self as Dictionary<TKey, TValue>)?.Comparer;
+        var copy = new Dictionary<TKey, ReadOnlyCollection<TValue>>(self.Count, keyComparer);
+        foreach (var kvp in self)
+        {
+            if (kvp.Value is null)
+            {
+                copy.Add(kvp.Key, null);
+                continue;
+            }
+
+            var readOnlyValue = kvp.Value as ReadOnlyCollection<TValue> ?? new ReadOnlyCollection<TValue>(kvp.Value);
+            copy.Add(kvp.Key, readOnlyValue);
+        }
+
+        return new ReadOnlyDictionary<TKey, ReadOnlyCollection<TValue>>(copy);
+    }
+}",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
         [Fact]
         public async Task NullCheck_AfterTryCast_Diagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -3337,6 +3337,29 @@ End Class");
 
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
+        [Fact, WorkItem(4411, "https://github.com/dotnet/roslyn-analyzers/issues/4411")]
+        public async Task NullCheck_AfterTryCast_03_NoDiagnostic()
+        {
+            await VerifyCSharpAnalyzerAsync(@"
+class A
+{
+    void M(object o)
+    {
+        if (o is A a)
+        {
+            _ = (a as B)?.ToString();
+        }
+    }
+}
+
+class B : A
+{
+}
+");
+        }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
         [Fact]
         public async Task NullCheck_AfterTryCast_Diagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -3360,7 +3360,7 @@ class B : A
 
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
-        [Fact, WorkItem(4411, "https://github.com/dotnet/roslyn-analyzers/issues/4411")]
+        [Fact, WorkItem(4383, "https://github.com/dotnet/roslyn-analyzers/issues/4383")]
         public async Task NullCheck_AfterTryCast_04_NoDiagnostic()
         {
             await new VerifyCS.Test

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -5476,6 +5476,34 @@ public class C : IDisposable
 }");
         }
 
+#if NETCOREAPP
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
+        [Fact, WorkItem(4327, "https://github.com/dotnet/roslyn-analyzers/issues/4327")]
+        public async Task TestCompilerGeneratedNullCheckNotFlagged_02()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class Class1
+{
+    static async IAsyncEnumerable<int> M(IAsyncEnumerable<int> iae)
+    {
+        await foreach (int i in iae.ConfigureAwait(false))
+        {
+            if (i > 0)
+                yield return i;
+        }
+    }
+}",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
+#endif
+
         [Fact]
         public async Task StaticObjectReferenceEquals_Diagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -5504,6 +5504,26 @@ class Class1
         }
 #endif
 
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
+        [Fact, WorkItem(4382, "https://github.com/dotnet/roslyn-analyzers/issues/4382")]
+        public async Task TestCompilerGeneratedNullCheckNotFlagged_03()
+        {
+            await VerifyCSharpAnalyzerAsync(@"
+using System;
+
+class C
+{
+    void F(IDisposable x)
+    {
+        if (x != null)
+        {
+            using (x) { }
+        }
+    }
+}");
+        }
+
         [Fact]
         public async Task StaticObjectReferenceEquals_Diagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -4215,6 +4215,27 @@ class Test
 
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
+        [Fact, WorkItem(4415, "https://github.com/dotnet/roslyn-analyzers/issues/4415")]
+        public async Task NullCheck_AfterDirectCast_NullableValueType_NoDiagnostic()
+        {
+            await VerifyCSharpAnalyzerAsync(@"
+using System;
+
+internal class Class1
+{
+    public static void M(object obj)
+    {
+        var d = (DateTime?)obj;
+        if (d != null)
+        {
+        }
+    }
+}
+");
+        }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.CopyAnalysis)]
         [Fact]
         public async Task NullCheck_AfterDirectCast_Diagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
@@ -3092,5 +3092,32 @@ public class Test
                 LanguageVersion = CSharpLanguageVersion.CSharp9
             }.RunAsync();
         }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.ValueContentAnalysis)]
+        [Fact, WorkItem(4387, "https://github.com/dotnet/roslyn-analyzers/issues/4387")]
+        public async Task RangeAndIndexOperation_NoDiagnostic()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+internal class Class1
+{
+    private static bool TryParseUnit(string unit)
+    {
+        char last = unit[^1];
+        if (last != 'b')
+            return false;
+
+        string subUnit = unit[1..];
+        if (subUnit != ""b"")
+            return false;
+
+        return true;
+    }
+}
+",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
     }
 }

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
@@ -3093,6 +3093,7 @@ public class Test
             }.RunAsync();
         }
 
+#if NETCOREAPP
         [Trait(Traits.DataflowAnalysis, Traits.Dataflow.ValueContentAnalysis)]
         [Fact, WorkItem(4387, "https://github.com/dotnet/roslyn-analyzers/issues/4387")]
         public async Task RangeAndIndexOperation_NoDiagnostic()
@@ -3119,5 +3120,6 @@ internal class Class1
                 LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
             }.RunAsync();
         }
+#endif
     }
 }

--- a/src/Utilities/FlowAnalysis/Extensions/BasicBlockExtensions.cs
+++ b/src/Utilities/FlowAnalysis/Extensions/BasicBlockExtensions.cs
@@ -137,6 +137,27 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis
             return false;
         }
 
+        /// <summary>
+        /// Returns true if the given basic block is the first block of a compiler generated finally region.
+        /// </summary>
+        public static bool IsFirstBlockOfCompilerGeneratedFinally(this BasicBlock basicBlock, ControlFlowGraph cfg)
+        {
+            if (!basicBlock.IsFirstBlockOfRegionKind(ControlFlowRegionKind.Finally, out var finallyRegion))
+            {
+                return false;
+            }
+
+            foreach (var operation in finallyRegion.DescendantOperations(cfg))
+            {
+                if (!operation.IsImplicit)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
         internal static ControlFlowRegion? GetInnermostRegionStartedByBlock(this BasicBlock basicBlock, ControlFlowRegionKind regionKind)
         {
             if (basicBlock.EnclosingRegion?.FirstBlockOrdinal != basicBlock.Ordinal)

--- a/src/Utilities/FlowAnalysis/Extensions/ControlFlowRegionExtensions.cs
+++ b/src/Utilities/FlowAnalysis/Extensions/ControlFlowRegionExtensions.cs
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System.Collections.Generic;
+
 namespace Microsoft.CodeAnalysis.FlowAnalysis
 {
     internal static class ControlFlowRegionExtensions
@@ -7,5 +9,17 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis
         public static bool ContainsRegionOrSelf(this ControlFlowRegion controlFlowRegion, ControlFlowRegion nestedRegion)
             => controlFlowRegion.FirstBlockOrdinal <= nestedRegion.FirstBlockOrdinal &&
             controlFlowRegion.LastBlockOrdinal >= nestedRegion.LastBlockOrdinal;
+
+        public static IEnumerable<IOperation> DescendantOperations(this ControlFlowRegion controlFlowRegion, ControlFlowGraph cfg)
+        {
+            for (var i = controlFlowRegion.FirstBlockOrdinal; i <= controlFlowRegion.LastBlockOrdinal; i++)
+            {
+                var block = cfg.Blocks[i];
+                foreach (var operation in block.DescendantOperations())
+                {
+                    yield return operation;
+                }
+            }
+        }
     }
 }

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
@@ -1156,7 +1156,7 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
                     var location = AbstractLocation.CreateAllocationLocation(operation, operation.Type, DataFlowAnalysisContext);
                     return PointsToAbstractValue.Create(location, mayBeNull: false);
                 }
-                else if (inference.IsUnboxing)
+                else if (inference.IsUnboxing && operation.Type.IsNonNullableValueType())
                 {
                     return PointsToAbstractValue.NoLocation;
                 }

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAnalysis.ValueContentDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/ValueContentAnalysis/ValueContentAnalysis.ValueContentDataFlowOperationVisitor.cs
@@ -177,7 +177,9 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.ValueContentAnalysis
                 _ = base.DefaultVisit(operation, argument);
                 if (operation.Type == null)
                 {
-                    return ValueContentAbstractValue.ContainsNullLiteralState;
+                    return operation.Kind == OperationKind.None ?
+                        ValueContentAbstractValue.MayBeContainsNonLiteralState :
+                        ValueContentAbstractValue.ContainsNullLiteralState;
                 }
 
                 if (ValueContentAbstractValue.IsSupportedType(operation.Type, out var valueTypeSymbol))

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntityFactory.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntityFactory.cs
@@ -214,6 +214,17 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
                         !_captureIdCopyValueMap.ContainsKey(flowCapture.Id) &&
                         analysisEntity.Type.IsValueType == capturedEntity.Type.IsValueType)
                     {
+                        // Skip flow capture for conversions unless we know the points to value
+                        // for conversion and operand is identical.
+                        if (flowCapture.Value is IConversionOperation conversion)
+                        {
+                            if (_getPointsToAbstractValue == null ||
+                                _getPointsToAbstractValue(conversion) != _getPointsToAbstractValue(conversion.Operand))
+                            {
+                                break;
+                            }
+                        }
+
                         var kind = capturedEntity.Type.IsValueType ? CopyAbstractValueKind.KnownValueCopy : CopyAbstractValueKind.KnownReferenceCopy;
                         var copyValue = new CopyAbstractValue(ImmutableHashSet.Create(analysisEntity, capturedEntity), kind);
                         _captureIdCopyValueMap.Add(flowCapture.Id, copyValue);

